src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Regex.Cache.cs

```diff
@@ -24,71 +28,92 @@ public static int CacheSize
                 if (value < 0)
                     throw new ArgumentOutOfRangeException(nameof(value));
 
-                s_cacheSize = value;
-                if (s_livecode.Count > s_cacheSize)
+                lock (s_livecode)
                 {
-                    lock (s_livecode)
+                    s_cacheSize = value;  // not to allow other thread to change it while we use cache
+                    while (s_livecode.Count > s_cacheSize)
                     {
-                        while (s_livecode.Count > s_cacheSize)
-                            s_livecode.RemoveLast();
+                        CachedCodeEntry last = s_livecode_last;
+                        s_livecode.Remove(last.Key);
+
+                        // update linked list:
+                        s_livecode_last = last.Next;
+                        if (last.Next != null)
+                            last.Next.Previous = null;
```

Comment:
Some additional asserts to help ensure the integrity of the list would be helpful, e.g.
```C#
if (last.Next != null)
{
    Debug.Assert(s_livecode_first != null && s_livecode_first != last);
    last.Next.Previous = null;
}
else
{
    Debug.Assert(s_livecode_first == last);
    Debug.Assert(s_livecode_last == null);
    s_livecode_first = null;
}
```
etc.

Reply:
Yes, good idea